### PR TITLE
C/C++ Interface

### PR DESCRIPTION
**Pull request by cwfitzgerald**
As this one is a bit of a chonker, I'm publishing this WIP as I make the final touchups to the documentation with higher level "how this api is structured" docs.

This adds the C api for bve, as well as the infrastructure needed to continue growing it.

C++, C#, etc wrappers are coming in another PR, as is documentation on the Rust members themselves.

### CODE REVIEW

File: bve-derive/src/lib.rs [hunk 1]
```diff
@@ -36,14 +40,18 @@
 #![allow(clippy::unreachable)]
 #![allow(clippy::wildcard_enum_match_arm)]
 
-extern crate proc_macro;
-
 use proc_macro::TokenStream;
 
-// Code coverage is silly on these derive
+// Code coverage is silly on these derives
```
Comment (DethRaid) [inline bve-derive/src/lib.rs]: you're silly on these derives

Comment (cwfitzgerald) [reply bve-derive/src/lib.rs]: Don't drink and derive

File: bve-native/src/lib.rs [hunk 2]
```diff
@@ -36,3 +51,151 @@
...
+use std::borrow::Cow;
+use std::ffi::{CStr, CString};
+use std::os::raw::*;
+use std::ptr::null_mut;
+
+/// C safe wrapper for an [`Option`].
+///
+/// # Safety
+///
+/// Reading from the `value` member is undefined behavior if `exists` is false. In practice it is zeroed.
```
Comment (DethRaid) [inline bve-native/src/lib.rs]: This is more information than I should be given

Comment (cwfitzgerald) [reply bve-native/src/lib.rs]: What do you want to be given?

Comment (DethRaid) [reply bve-native/src/lib.rs]: You just told me that this is `sizeof(T)` bytes of memory which I do not have to allocate and may use as I please

**Pull request merged**